pkg/store/tree.go

```diff
@@ -0,0 +1,151 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package store
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/onosproject/onos-config/pkg/store/change"
+	"strconv"
+	"strings"
+)
+
+// BuildTree is a function that takes an ordered array of ConfigValues and
+// produces a structured formatted JSON tree
+func BuildTree(values []*change.ConfigValue) ([]byte, error) {
+
+	root := make(map[string]interface{})
+	rootif := interface{}(root)
+	for _, cv := range values {
+		err := addPathToTree(cv.Path, cv.Value, &rootif)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	buf, err := json.Marshal(root)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf, nil
+}
+
+// addPathToTree is a recursive function that builds up a map
+// suitable for using with json.Marshal, which in turn can be used to feed in to
+// ygot.Unmarshall
+// This follows the approach in https://blog.golang.org/json-and-go "Generic JSON with interface{}"
+func addPathToTree(path, value string, nodeif *interface{}) error {
+	pathelems := strings.Split(path, "/")
+
+	// Convert to its real type
+	nodemap, ok := (*nodeif).(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("Could not convert nodeif %v for %s", *nodeif, path)
+	}
+
+	if len(pathelems) == 2 && value != "" {
+		// At the end of a line - this is the leaf
+		numval, err := strconv.Atoi(value)
+		if err == nil {
+			(nodemap)[pathelems[1]] = numval
+		} else if strings.ToLower(value) == "true" || strings.ToLower(value) == "false" {
+			(nodemap)[pathelems[1]] = strings.ToLower(value)
+		} else {
+			(nodemap)[pathelems[1]] = value
+		}
+	} else if strings.Contains(pathelems[1], "=") {
+		// To handle list index items
+		refinePath := strings.Join(pathelems[2:], "/")
+		if refinePath == "" {
+			return nil
+		}
+		refinePath = fmt.Sprintf("/%s", refinePath)
+		brktIdx := strings.Index(pathelems[1], "[")
+		eqIdx := strings.Index(pathelems[1], "=")
+		brktIdx2 := strings.Index(pathelems[1], "]")
+
+		listName := pathelems[1][:brktIdx]
+		keyName := pathelems[1][brktIdx+1 : eqIdx]
+		keyVal := pathelems[1][eqIdx+1 : brktIdx2]
+
+		listSlice, ok := nodemap[listName]
+		if !ok {
+			listItemMap := make(map[string]interface{})
+			listItemMap[keyName] = keyVal
+			listItemIf := interface{}(listItemMap)
+
+			err := addPathToTree(refinePath, value, &listItemIf)
+			if err != nil {
+				return err
+			}
+			listItemSlice := make([]interface{}, 1)
+			listItemSlice[0] = listItemMap
+
+			(nodemap)[listName] = listItemSlice
+		} else {
+			//Reuse existing listSlice
+			listSliceIf, ok := listSlice.([]interface{})
+			if !ok {
+				return fmt.Errorf("Failed to convert list slice %s", listName)
+			}
+			for idx, ls := range listSliceIf {
+				lsMap, ok := ls.(map[string]interface{})
+				if !ok {
+					return fmt.Errorf("Failed to convert list slice %d", idx)
+				}
+				if (lsMap)[keyName] == keyVal {
+					fmt.Println("Found", keyName, keyVal)
```

Comment:
is this just to print ? no assignment ?

Reply:
This has been fixed - added corresponding new use case with **rx-power**